### PR TITLE
9484 report tile perm optimize

### PR DESCRIPTION
**Pull request by whatisgalen**
### Types of changes

-   [ ] Bugfix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Description of Change

There are some examples where a list of tiles gets checked against a user's permissions. The issue (#9484)  is that we only care about whether the user has permission for the nodegroup, but we're checking all tiles and thus we're checking tile.nodegroup permissions even after we already know whether the user has permission for that nodegroup.

The two changes made in this PR are in the Resource proxy model to more efficiently leverage the permission backend in the `load_tiles` method, as well as in the Resource Report API to achieve the same for tiles and cards tied to the user's permissions for their related nodegroups.

### Issues Solved

#9484 

### Checklist

-   [ ] Unit tests pass locally with my changes
-   [ ] I have added tests that prove my fix is effective or that my feature works
-   [ ] I have added necessary documentation (if appropriate)

#### Ticket Background
*   Sponsored by: Los Angeles OHR
*   Found by: @whatisgalen 
*   Tested by: @ 
*   Designed by: @whatisgalen

### CODE REVIEW

**Review by jacobtylerwalls — CHANGES_REQUESTED** [review]
Thanks, I confirmed the reduction in queries.

The tiles change is a win all around. The cards change causes one new query, a Nodegroups.objects.all(), but it seems to always pay itself back with avoiding lots of user_permission_cache queries, so that's a win too.

Some little things and I'd be good with a merge.
***
If you're into it, I'm hoping to start writing tests for these kinds of changes with the `self.assertNumQueries()` helper so that we don't have to manually test so much. But we can leave that for the next one.

**Review by jacobtylerwalls — COMMENTED** [review]
Hey @whatisgalen :wave:

While here, for the second part of the PR with the permitted cards optimization, what do you think about attacking those trips to the db just to sort by cardorder?

I'm showing one saved query per permitted card with:

```diff
diff --git a/arches/app/views/api.py b/arches/app/views/api.py
index b8d2b9983..e32c44798 100644
--- a/arches/app/views/api.py
+++ b/arches/app/views/api.py
@@ -14,6 +14,7 @@ from rdflib import RDF
 from rdflib.namespace import SKOS, DCTERMS
 from revproxy.views import ProxyView
 from slugify import slugify
+from operator import attrgetter
 from urllib import parse
 from collections import OrderedDict
 from django.contrib.auth import authenticate
@@ -1115,14 +1116,18 @@ class ResourceReport(APIBase):
 
         if "cards" not in exclude:
             # collect the nodegroups for which this user has perm
-            readable_nodegroups = [str(nodegroup.pk) for nodegroup in get_nodegroups_by_perm(request.user, [perm], any_perm=True)]
+            readable_nodegroups = get_nodegroups_by_perm(request.user, [perm], any_perm=True)
             
             # query only the cards whose nodegroups are readable by user
-            permitted_cards = list(CardProxyModel.objects.filter(graph_id=resource.graph_id, nodegroup_id__in=readable_nodegroups).select_related("nodegroup").order_by("sortorder"))
+            permitted_cards = (
+                CardProxyModel.objects.filter(graph_id=resource.graph_id, nodegroup__in=readable_nodegroups)
+                .prefetch_related("cardxnodexwidget_set")
+                .order_by("sortorder")
+            )
 
             cardwidgets = [
                 widget
-                for widgets in [card.cardxnodexwidget_set.order_by("sortorder").all() for card in permitted_cards]
+                for widgets in [sorted(card.cardxnodexwidget_set.all(), key=attrgetter("sortorder")) for card in permitted_cards]
                 for widget in widgets
             ]
```

I also noticed the select_related(nodegroup) is unnecessary.

File: arches/app/views/api.py [hunk 1]
```diff
@@ -1108,20 +1108,17 @@ def get(self, request, resourceid):
...
-            permitted_cards = []
-            for card in CardProxyModel.objects.filter(graph_id=resource.graph_id).select_related("nodegroup").order_by("sortorder"):
-                if request.user.has_perm(perm, card.nodegroup):
-                    card.filter_by_perm(request.user, perm)
-                    permitted_cards.append(card)
+            # collect the nodegroups for which this user has perm
+            readable_nodegroups = [str(nodegroup.pk) for nodegroup in get_nodegroups_by_perm(request.user, [perm], any_perm=True)]
+            
+            # query only the cards whose nodegroups are readable by user
+            permitted_cards = list(CardProxyModel.objects.filter(graph_id=resource.graph_id, nodegroup_id__in=readable_nodegroups).select_related("nodegroup").order_by("sortorder"))
```
[the remark this answers is not in the document]
Comment (whatisgalen) [reply arches/app/views/api.py:1121]: Originally `permitted_cards` is a list, so I assumed it best to preserve that data type

**Comment by whatisgalen** [comment]
> +            readable_nodegroups = get_nodegroups_by_perm(request.user, [perm], any_perm=True)
>              
>              # query only the cards whose nodegroups are readable by user
> -            permitted_cards = list(CardProxyModel.objects.filter(graph_id=resource.graph_id, nodegroup_id__in=readable_nodegroups).select_related("nodegroup").order_by("sortorder"))
> +            permitted_cards = (
> +                CardProxyModel.objects.filter(graph_id=resource.graph_id, nodegroup__in=readable_nodegroups)
> +                .prefetch_related("cardxnodexwidget_set")
> +                .order_by("sortorder")
> +            )
>  
>              cardwidgets = [
>                  widget
> -                for widgets in [card.cardxnodexwidget_set.order_by("sortorder").all() for card in permitted_cards]
> +                for widgets in [sorted(card.cardxnodexwidget_set.all(), key=attrgetter("sortorder")) for card in permitted_cards]
>                  for widget in widgets
>              ]
> ```

What's up with using `sorted` in the widget list comp? If the intent is to speed optimize there, maybe we just use a set for the inner comp instead of a list and in that case sorting becomes unnecessary?

**Review by whatisgalen — COMMENTED** [review]
requested changes committed

**Review by jacobtylerwalls — APPROVED** [review]
Thanks for the updates!

**Pull request merged**